server/src/main/java/org/opensearch/common/network/NetworkService.java

```diff
@@ -237,6 +239,43 @@ public InetAddress resolvePublishHostAddresses(String publishHosts[]) throws IOE
         return addresses[0];
     }
 
+    /**
+     * Resolve the publishPort for a server provided a list of boundAddresses and a publishInetAddress.
+     * Resolution strategy is as follows:
+     * If a configured port exists resolve to that port.
+     * If a bound address matches the publishInetAddress resolve to that port.
+     * If a bound address is a wildcard address resolve to that port.
+     * If all bound addresses share the same port resolve to that port.
+     *
+     * @param publishPort -1 if no configured publish port exists
+     * @param boundAddresses addresses bound by the server
+     * @param publishInetAddress address published for the server
+     * @return Resolved port. If publishPort is negative and no port can be resolved return publishPort.
+     */
+    public static int resolvePublishPort(int publishPort, List<TransportAddress> boundAddresses, InetAddress publishInetAddress) {
```

Comment:
I think we could generalize the implementation with `TcpTransport.resolvePublishPort` under `Transport` interface?

Reply:
Replaced redundant `resolvePublishPort` with a helper in `TcpTransport`.